write dplyr.rmd completed

damianooldoni opened the pull request:
I wrote the tutorial dplyr.rmd. Reviews are more than welcome!
I knitted it too to check the result. How avoid starting chapters count from 0?
Thanks.

damianooldoni committed:
Read, explore, sort, select chapters

damianooldoni committed:
Correct order of exercises first part

damianooldoni committed:
add two chapters: pipes and distinct

damianooldoni committed:
add group_by summarise, spread, gather

damianooldoni committed:
add rename and mutate

damianooldoni committed:
Do not overwrite data in mutate exercises

damianooldoni committed:
Add chapter about recode and case when

damianooldoni committed:
add export chapter

damianooldoni committed:
review content and Rmd options

damianooldoni committed:
Tune up the html ouput

stijnvanhoey commented on docs/dplyr.Rmd:
```diff
@@ -0,0 +1,447 @@
+* **col_names(df)** display column names
+* **nrow(df)** returns the number of rows of the dataframe
+* **ncol(df)** return the number of rows of the dataframe
+* **mean(vector)** returns the (arithmetic) mean
```
this is maybe tricky. Do we expect participants to understand all the data types and how to distinguish a `vector` and a `df`? Why not just apply it to `df`?

damianooldoni replied:
The problem of applying `mean()`to the df is that you get an error!  
> Warning message:
In mean.default(a) : argument is not numeric or logical: returning NA

And actually this error is not due by the fact that some columns are not numeric (e.g. `date`and `species`). I have just tried `mean()` with a numeric dataframe and I got the same error... There several ways to calculate the mean:
```r
library(purrr)
means <- map_dbl(m, mean)
```

or

```r
sapply(data, mean, na.rm = T)
```

I find them not "easy enough" to be insert in this tutorial at least not at the very beginning. If we suppose the participants are a little familiar with R but not with dplyr, then i can conclude they know how to use the `$` to extract a column as vector, or am I wrong?

damianooldoni replied:
However, `min()` and `max()` work well on a all-numeric dataframe! Thus, I remove `mean()` from the list: no distinction dataframe-vector anymore! Solved here: ea9ff1b7d0c6f9e371e049609c90eb59aae9ea5a

stijnvanhoey commented on docs/dplyr.Rmd:
```diff
@@ -0,0 +1,447 @@
+
+* **view(df)** opens the entire dataframe in viewer
+* **head(df)** returns the first part of a data frame. It works with several other data types (vectors, tables, ...). Specifiy how many rows to display with parameter `n`: `haed(df, n = )`, default `n = 6`
+* **str(df)** displays the internal structure of the dataframe
```
isn't this a function for a more advanced user? It provides very technical information... maybe focus on `class`, which provides the info: 'yes, it is a data.frame'.

damianooldoni replied:
Nice: `class(df)` added, indeed. However, I did not remove `str(df)`: it is just a list of tools, they use what they feel more useful. Implemented here: ea9ff1b7d0c6f9e371e049609c90eb59aae9ea5a

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+
+7. What is the mean weight?
+```{r}
+mean(data$weight_in_g, na.rm = TRUE)
````
Do we expect the users to know how to use the `$`-sign? 

Furthermore, is this part of the course or do you want to go for `dplyr` all the way, like:
```
data %>% summarize(mean_weight = mean(weight_in_g, na.rm = TRUE))
```

damianooldoni replied:
By removing `mean()` from the list as explained before, this exercise doesn't suit this chapter anymore. Furthermore, you are right: working with dplyr means using summarise to do it, so I move it to the chapter about `summarize()` :+1: See commit here: c166ad908b395dce81ce73d89c6c8af958d421ff

stijnvanhoey commented on docs/dplyr.Rmd:
```diff
@@ -0,0 +1,447 @@
+
+## Select columns and rows
+You can select specific columns of a dataframe by `select()`. To select rows use `filter()`. Only columns/rows where conditions are true are selected. For example, selecting a column by its column name:  `select(df, column_name)`. To select everything except a specific column use the minus symbol before its name: `select(df, -column_name)`.
+
```
Maybe a single example would be helpful for participants to get started, now it is directly going to exercises...

damianooldoni replied:
Completely right! I implemented it everywhere actually by using a dataframe, `CO2`, which is one of many dataframes R imports automatically when starting. See commit 4df18c9404ddbdb6752770d0a3835c1081c39f77

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+```
+
+## Pipes
+As you probably noticed, we need to create new (temporary) variables everytime we want to apply two or more functions. This method has several disadvantages, i.a. messy workspace, high memory space usage, and higher chance of typos.
````
I am not completely sure about `high memory space usage`... Is that so? Is `dplyr` doing each time a deep copy and reallocating new memory? I don't think so...

damianooldoni replied:
Actually I want to say that dplyr is better because you don't need to save so many variables. I should remove temporary then, yes. Done: 5a548820ed0a0256cc28d71311267e9d783435fc

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+```
+
+This way of working with functions is typical of programming and is implemented in many other programming languages. The disadvantage of this approach is readibility: you have to read from right to left in order to understand what happens with our data!
+
````
I like this! cool intro

damianooldoni replied:
thanks

stijnvanhoey commented on docs/dplyr.Rmd:
```diff
@@ -0,0 +1,447 @@
+| trousers | clothes  | 92    |
+| t-shirt  | clothes  | 25    |
+| peers    | fruit    |  5    |
+
```
I would introduce it like it is done in carpentry: 
- first: you want to calculate sum/mean/...  => do it on df: `df %>% summarise(sum = sum(price))`  (cfr. earlier remark)
- then: but you have groups defined by other columns (factors) and you now want now to do the same for each of these groups... Well, but `groupby` in between...

stijnvanhoey replied:
When I say *(cfr. earlier remark)*, it is the inconsistency with the first batch of exercises where mean is derived differently. As such, `summarise` seems something interlinked with `groupby` after reading the next section. But, the groupby concept can be used for other cases as well...

damianooldoni replied:
Right! Fixed here 4df18c9404ddbdb6752770d0a3835c1081c39f77

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+
+`group_by` actually doesn't modify the data contained in the dataframe, but adds an internal grouping to the dataframe, which is visible if you run:
+```r
+class(df%>% group_by(category))
````
cfr. earlier comment about `class` or `str`. When used here, also make sure `class` is introduced in the beginning. I would remove `str` (intro and exercise)

damianooldoni replied:
`class`introduced in the beginning among other exploratory variables.

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+2. Show mean weight per sex
+```{r group_by & summarise()}
+data %>% filter(!is.na(weight_in_g)) %>% 
+  group_by(sex) %>% summarise(sum = sum(weight_in_g))
````
`sum` -> `mean` (I'll add in PR). 
Why suddenly here with `filter` and not mean(..., na.rm = TRUE)?

damianooldoni replied:
You are right! I have corrected it in 0312d0b5f1c4a33675d85037cfa6dbef937d4e81
```r
data %>%
  group_by(sex) %>% 
  summarize(mean_weight_in_g = mean(weight_in_g, na.rm = TRUE))
```

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+```
+
+2. Gather back the data and save output as `weight_per_species_sex_gather`. Is `weight_per_species_sex_gather` equal to `weight_per_species_sex`?
+```{r}
````
Maybe add a test here to effectively check it?

damianooldoni replied:
It is an open question because you get the same dataframe but the columns are not in the same order. Maybe i should add a line or tow of code where I show the column names (`colnames()`) and see that they are not in the same order... You are right.

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+recode(x, a = "Apple")
+#>  [1] "b"     "b"     "Apple" "c"     "c"     "Apple" "c"     "c"     "Apple"
+#> [10] "c"
+```
````
I would provide an example applied on a data.frame...

damianooldoni replied:
The point is that `recode()` doesn't work with df:
Checked with this code:
```r
a <- data.frame(b = c(3,2), c = c(4,5))
dplyr::recode(a, b = c(4,5))
```
Got: 
*Error in UseMethod("recode") : 
  no applicable method for 'recode' applied to an object of class "data.frame"*

stijnvanhoey replied:
so, people won'tbe able to use it anyway when focusing on tidyverse/dplyr? maybe just provide only `case_when`?

peterdesmet replied:
We use recode in a mutate(), as we do for case_when()

damianooldoni replied:
I leave `recode` functionality as it is. I have just added an example based on the CO2 dataframe as almost in every chapter See 4df18c9404ddbdb6752770d0a3835c1081c39f77

stijnvanhoey commented on docs/dplyr.Rmd:
````diff
@@ -0,0 +1,447 @@
+After hours of data analysis maybe you would like to save the output, isn't it?
+ You can save a dataframe object by means of function `write_table()`:
+ ```r
+ write.table(df, file = "./save/it/here/df.csv", ...)
````
`write_table` is a utils-package function. Should we opt for `tidyverse`-packages and use the `readr`-package alternatives?

damianooldoni replied:
You are right. I changed `write.table()` to `write_tsv()` and `write_csv()` from readr package. 
See here: 49c535c8b205d6dda1345dc9ff60b6a01180224d and here e9a9d9c60df889cecce581b971677e16670db8a7

peterdesmet commented:
Regarding section numbering: start your titles at level H1, so `## Read data` should be `# Read data` and all further titles should have one `#` less. I think that should fix it.

Also:

* Add a empty line after each heading, paragraph, code block and list (not list item). This will be interpreted correctly by any markdown parser and it is easier to read in it's row form
* Place literal things the user has to see, click, etc. in `backticks`

And personal preferences:

* If you're listing sequential steps, it's easier to number them. You can either opt for 1. 1. 1. or 1. 2. 3. Both will render as 1. 2. 3., but I prefer to use 1. 2. 3. so it makes sense in its raw form
* Start step list items with a capital (because they might contain multiple sentences)
* Don't end list items with comma or dot (but sentences in the middle get one... like my notes here)

From this:

```markdown
## Read data
First of all, we need to import data in R. The most user friendly way to do it is by using the interface provided by RStudio. In the Environment window, click on *Import Dataset* and select *From Text (readr)...*. A window will pop-up. Follow the next steps:
- click on *Browse* to selet the text file you want to import,
- select the right delimiter,
- check whether the class of the column values fits your purpose and in case apply changes by clicking on the class name below the column name,
- change variable name in *code preview* if you prefer to save the data with a different name than filename,
- click on *Import*
A dataframe is automatically generated and, very handy, the code to generate it too!

### Exercise
1. Read `survey_data_spreadsheet_tidy.csv`. Check whether the columns match the following types
```

To this:

```markdown
# Read data

First of all, we need to import data in R. The most user friendly way to do it is by using the interface provided by RStudio. In the Environment window, click on `Import Dataset` and select `From Text (readr)...`. A window will pop-up. Follow the next steps:

1. Click on `Browse` to select the text file you want to import
2. Select the right delimiter
3. Check whether the class of the column values fits your purpose and in case apply changes by clicking on the class name below the column name
4. Change variable name in `code preview` if you prefer to save the data with a different name than filename
5. Click on `Import`

A dataframe is automatically generated and, very handy, the code to generate it too!

## Exercise

1. Read `survey_data_spreadsheet_tidy.csv`. Check whether the columns match the following types
```

damianooldoni committed:
move dplyr.Rmd from /docs to /src

damianooldoni committed:
increase readibility

Based on review of @peterdesmet: insert spaces, remove one hashtag ("#") everywhere

damianooldoni committed:
remove mean() and improve readibility list exploratory functions

damianooldoni committed:
improve readibility

Following suggestions of @peterdesmet

damianooldoni committed:
add read_csv2() as alternative solution

add read_csv2() as alternative solution to import semicolon separated text files

damianooldoni committed:
add encouragement to code insteda of suing always the GUI

Do not be lazy! Learn coding by GUI should be the message!

damianooldoni committed:
move mean weight exercise to chapter about summarise

as consequence of a right comment of @stijnvanhoey and the fact that I removed mean() from the list of exploratory functions for dfs

peterdesmet commented:
As this has already been through review, I'm going to merge this into master right away, so it can be integrated with bookdown. I might review it further from there.
